### PR TITLE
Scale down constraint evaluation API proposal

### PR DESCRIPTION
**Pull request by tbak**

### CODE REVIEW

File: fenzo-core/src/main/java/com/netflix/fenzo/ScaleDownConstraintExecutor.java [hunk 1]
```diff
@@ -1,38 +1,92 @@
...
+            Optional<? super Object> optionalContext = Optional.empty();
+            for (VirtualMachineLease lease : groupCandidates) {
+                double currentScore = scores.getOrDefault(lease, 0.0);
+                if (currentScore != NOT_REMOVABLE_MARKER) {
+                    ScaleDownConstraintEvaluator.Result result = e.evaluate(lease, optionalContext);
+                    double newScore = result.getScore();
+                    if (newScore == 0) {
+                        scores.put(lease, NOT_REMOVABLE_MARKER);
+                    } else {
+                        scores.put(lease, currentScore + newScore);
```
Comment (spodila) [inline fenzo-core/src/main/java/com/netflix/fenzo/ScaleDownConstraintExecutor.java]: This looks like all evaluators are treated equal. We wanted to have weights for combining the scores. Wouldn't that be done here?

File: fenzo-core/src/main/java/com/netflix/fenzo/ScaleDownConstraintExecutor.java [hunk 2]
```diff
@@ -1,38 +1,92 @@
...
+
+        if (logger.isDebugEnabled()) {
+            List<String> hosts = scaleDownOrder.stream().map(VirtualMachineLease::hostname).collect(Collectors.toList());
+            logger.debug("Evaluated scale down order: {}", hosts);
+        }
+
+        return scaleDownOrder;
+    }
+
+    private Stream<VirtualMachineLease> groupEvaluator(Set<VirtualMachineLease> groupCandidates) {
```
Comment (spodila) [inline fenzo-core/src/main/java/com/netflix/fenzo/ScaleDownConstraintExecutor.java]: Would be good to have a test that exercises multiple evaluators with weighted scores.

File: fenzo-core/src/main/java/com/netflix/fenzo/ScaleDownConstraintExecutor.java (hunk 1 above)
Comment (tbak) [inline fenzo-core/src/main/java/com/netflix/fenzo/ScaleDownConstraintExecutor.java]: Yes, there are no weight there right now. I assumed they are all equal, but I agree adding weights makes sense, and is not complicated. I will change that.

File: fenzo-core/src/main/java/com/netflix/fenzo/ScaleDownConstraintExecutor.java (hunk 2 above)
Comment (tbak) [inline fenzo-core/src/main/java/com/netflix/fenzo/ScaleDownConstraintExecutor.java]: I have added weights + corresponding test in my last PR.

**Pull request merged**